src/posts/gpu-pipeline/index.md

```diff
@@ -0,0 +1,223 @@
+
+![baseline plot](baseline.png)
+
+In the plot above, we show the throughput of the data loading and training steps in our pipeline. The three bars represent:
+
+- Real Data: Baseline throughput of the end-to-end pipeline using real data.
+- No Training (i.e. data loading throughput): Throughput of the data loading without any training (to measure the time spent on data loading vs. training).
+- Synthetic Data (i.e. Training throughput): Throughput of the data loading using synthetic data (to remove the data loading bottleneck).
+
+The results show that the data loading step is the main bottleneck in our pipeline, with **much** lower throughput compared to the training step.
+
+## Hackathon: Putting this altogether
+
+Our initial profiling showed that data loading is a major bottleneck in this workflow.
+
+During the hackathon, we tested the following strategies to improve the data loading performance:
+
+1. Optimized Chunking & Compression
+   - We explored different chunking and compression strategies to optimize the data loading performance. We found that using Zarr v3 with optimized chunking and compression significantly improved the data loading performance.
+2. GPU native data loading with Zarr V3 and KvikIO
+3. Using `nvcomp` for decompression on GPUs
+4. NVIDIA DALI: We explored integrating NVIDIA's Data Loading Library (DALI) into Xarray to facilitate efficient data loading and preprocessing directly on the GPU. DALI provides highly optimized building blocks and an execution engine for data processing, accelerating deep learning applications.
+
+### Step 1: Optimized chunking :card_file_box:
+
+The ERA-5 dataset we were using had a sub-optimal chunking scheme of `{'time': 10, 'channel': C, 'height': H, 'width': W}`, which meant that a minimum of 10 timesteps of data was being read even if we only needed 2 consecutive timesteps at a time.
+We decided to rechunk the data to align with our access pattern of 1-timestep at a time, while reformating to Zarr v3.
+The full script is available [here](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/blob/main/rechunk/era5_rechunking.ipynb), with the main code looking like so:
+
+```python
+import xarray as xr
+
+ds: xr.Dataset = xr.open_mfdataset("ERA5.zarr")
+# Rechunk the data
+ds = ds.chunk({"time": 1, "level": 1, "latitude": 640, "longitude": 1280})
+# Save to Zarr v3
+ds.to_zarr("rechunked_ERA5.zarr", zarr_version=3)
+```
+
+For more optimal performance, consider:
+
+1. Storing the data without compression (if not transferring over a network), as decompressing data can slow down read speeds. But see also GPU decompression with nvCOMP below. :wink:
+2. Concatenating several data variables together **if** a single chunk size is too small (<1MB), at the expense of reducing readability of the Zarr store. Having too many small chunks can be detrimental to read speeds. A compressed chunk should be >1MB, <10MB (??TODO verify) for optimal reads.
```

Comment:
Need to escape the "<" character here to fix build error "Unexpected character `1` (U+0031) before name, expected a character that can start a name, such as a letter, `$`, or `_`".

```suggestion
2. Concatenating several data variables together **if** a single chunk size is too small (\<1MB), at the expense of reducing readability of the Zarr store. Having too many small chunks can be detrimental to read speeds. A compressed chunk should be >1MB, \<10MB (??TODO verify) for optimal reads.
```

Also, reminder that we need to verify if these chunk sizes (1MB, 10MB) are correct, my memory is a bit foggy now :face_in_clouds:

Reply:
Thanks! I spent some time trying to figure what went wrong in my local deployment and then I say your comment. 😉

Reply:
I changed this to be between 1MB and 100MB. 5-100 MB per chunk is along the line of what we recommend for lustre. Our chunk sizes were around 6.25 MiB.